experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts

```diff
@@ -131,120 +139,131 @@ export const optionalChangeRebaser: FieldChangeRebaser<OptionalChangeset> = {
 			}
+			const withIntention = (id: RegisterId): RegisterId => {
+				if (id === "self") {
+					return id;
+				}
+				const intention = getIntention(id.revision ?? revision, revisionMetadata);
+				return { revision: intention, localId: id.localId };
+			};
 
-			if (change.fieldChange !== undefined) {
-				if (fieldChange === undefined) {
-					fieldChange = {
-						id: change.fieldChange.id,
-						revision: fieldChangeInfo?.revision,
-						wasEmpty: change.fieldChange.wasEmpty,
-					};
-				} else {
-					fieldChange.id = change.fieldChange.id;
-					fieldChange.revision = fieldChangeInfo?.revision;
+			if (change.build !== undefined) {
+				for (const { id, set } of change.build) {
+					builds.push({
+						id: {
+							revision: getIntention(id.revision ?? revision, revisionMetadata),
+							localId: id.localId,
+						},
+						set,
+					});
 				}
+			}
+			const nextSrcToDst = new ChildChangeMap<
+				[dst: RegisterId, target: "nodeTargeting" | "cellTargeting"]
+			>();
+			const nextDstToSrc = new ChildChangeMap<RegisterId>();
 
-				let hasMatchingPriorInverse = false;
-				const maybePriorInverse = changes.findIndex((c) => {
-					const cChangeInfo = revisionMetadata.tryGetInfo(
-						// Change c may be a composite, in which case we need to look the revision of the fieldChange
-						c.revision ?? c.change.fieldChange?.revision,
-					);
-					return (
-						(cChangeInfo?.rollbackOf !== undefined &&
-							cChangeInfo?.rollbackOf === fieldChangeInfo?.revision) ||
-						(cChangeInfo?.revision !== undefined &&
-							cChangeInfo?.revision === fieldChangeInfo?.rollbackOf)
+			if (change.reservedDetachId !== undefined) {
+				latestReservedDetachId = withIntention(change.reservedDetachId);
+			}
+
+			// Compose all the things that `change` moved.
+			for (const [src, dst, target] of change.moves) {
+				let originalSrc = current.dstToSrc.get(withIntention(src));
+				let currentTarget: "cellTargeting" | "nodeTargeting" = "cellTargeting";
+				if (originalSrc !== undefined) {
+					const [dst2, existingTarget] =
+						current.srcToDst.get(originalSrc) ?? fail("expected backward mapping");
+					assert(
+						areEqualRegisterIds(dst2, withIntention(src)),
+						"expected consistent backward mapping",
 					);
-				});
-				hasMatchingPriorInverse = maybePriorInverse !== -1 && maybePriorInverse < index;
-
-				if (change.fieldChange.newContent !== undefined) {
-					if (hasMatchingPriorInverse) {
-						fieldChange = undefined;
-					} else {
-						fieldChange.newContent = { ...change.fieldChange.newContent };
-					}
+					currentTarget = existingTarget;
 				} else {
-					if (hasMatchingPriorInverse) {
-						fieldChange = undefined;
-					} else {
-						delete fieldChange.newContent;
-					}
+					originalSrc = withIntention(src);
 				}
+				nextSrcToDst.set(originalSrc, [
+					withIntention(dst),
+					target === "nodeTargeting" || currentTarget === "nodeTargeting"
+						? "nodeTargeting"
+						: "cellTargeting",
```

Comment:
It's hard to look at this and grok what it means for all 6 combinations.

I think one little thing that would have helped me woud be to model the case where there is no `originalSrc` with an undefined `currentTarget` (instead of setting it to `cellTargeting`, which is arbitrary).

Then perhaps this expression becomes easier to read:
```suggestion
					target === "cellTargeting" && currentTarget === "cellTargeting"
						? "cellTargeting"
						: "nodeTargeting",
```

Reply:
I don't think that expression works--when processing a move that has no `originalSrc` (i.e. the move's pre-existing source *is* its original source), we'll have `currentTarget` undefined and `target` is `cellTargeting`, but end up with `nodeTargeting`. So e.g. `compose({ moves: [["self", { localId: 1 }, "cellTargeting"]] })` would suddenly become `nodeTargeting`.

That said, I see your point about initializing to undefined making the decision point more uniform. I'll do that.

Reply:
Ah good point, thanks.